src/html/Quoting.cc

```diff
@@ -9,50 +9,48 @@
 #include "squid.h"
 #include "html/Quoting.h"
 
-/*
- *  HTML defines these characters as special entities that should be quoted.
- */
-static struct {
-    unsigned char code;
-    const char *quote;
-} htmlstandardentities[] =
+#include <array>
+#include <cstring>
+
 
+static const auto& MakeEscapeSequences()
 {
-    /* NOTE: The quoted form MUST not be larger than 6 character.
-     * see close to the MemPool commend below
+    static std::array<std::pair<unsigned char, const char *>, 5> const escapePairs = {
+        std::make_pair('<', "&lt;"),
+        std::make_pair('>', "&gt;"),
+        std::make_pair('"', "&quot;"),
+        std::make_pair('&', "&amp;"),
+        std::make_pair('\'', "&apos;")
+    };
+    static auto escapeMap = new std::array<const char *, 256>{};
+    const size_t maxEscapeLength = 7;
+    /* Encode control chars just to be on the safe side, and make
+     * sure all 8-bit characters are encoded to protect from buggy
+     * clients
      */
-    {
-        '<', "&lt;"
-    },
-    {
-        '>', "&gt;"
-    },
-    {
-        '"', "&quot;"
-    },
-    {
-        '&', "&amp;"
-    },
-    {
-        '\'', "&#39;"
-    },
-    {
-        0, NULL
+    for (uint32_t ch = 0; ch < 256; ++ch) {
+        if ((ch <= 0x1F || ch >= 0x7f) && ch != '\n' && ch != '\r' && ch != '\t') {
+            (*escapeMap)[ch] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+            snprintf(const_cast<char*>((*escapeMap)[ch]), sizeof escapeMap[ch], "&#%d;", static_cast<int>(ch));
+        }
+    }
+    for (auto &pair: escapePairs) {
+        (*escapeMap)[pair.first] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+        xstrncpy(const_cast<char*>((*escapeMap)[pair.first]), pair.second, maxEscapeLength);
     }
-};
+    return *escapeMap;
+}
+
 
-/*
- *  html_do_quote - Returns a static buffer containing the quoted
- *  string.
- */
 char *
 html_quote(const char *string)
 {
-    static char *buf = nullptr;
+    static const auto htmlSpecialCharacters = MakeEscapeSequences();
```

Comment:
We do not want to copy this array, even after migration to SBuf:

```suggestion
    static const auto &htmlSpecialCharacters = MakeEscapeSequences();
```

I also recommend renaming htmlSpecialCharacters variable to `escapeSequences` because it does not actually contain or supply special characters in any applicable sense, but I do not insist on that naming fix.

```suggestion
    static const auto &escapeSequences = MakeEscapeSequences();
```

Reply:
done in 446feb58d9